### PR TITLE
Dashboard user setting

### PR DESCRIPTION
**Pull request by bhatiadheeraj**

### CODE REVIEW

File: ui/src/settings.vue [hunk 1]
```diff
@@ -668,10 +733,54 @@ export default {
...
+                this.userEdit.scopes[key].forEach(permission=>{
+                    console.log(key,permission);
+                    this.scopeModel[key][permission] = true
+                });
+            }
             this.openids = user.ext.openids[0] || " ";
+            this.showModal = true;
+        },
+        checkValidJson(item) {
+            item = typeof item !== "string"
```
Comment (soichih) [inline ui/src/settings.vue:769]: I think item should always be string, no?

File: ui/src/settings.vue [hunk 2]
```diff
@@ -684,11 +793,17 @@ export default {
         },
         submitUser(e) {
             e.preventDefault();
-            this.userEdit.profile = JSON.parse(this.profile||"{}");
-            this.userEdit.scopes = JSON.parse(this.scopes||"{}");
+            if(!this.checkValidJson(this.userEditJSON)) {
+                this.$notify({type: "error", text: "Json formatting Error"});
+                return;
+            }
+            Object.keys(this.scopeModel).forEach(entry => {
```
Comment (soichih) [inline ui/src/settings.vue:800]: I think it's clearner to do
```
for(const key in this.scopeModel) {
```

File: ui/src/settings.vue [hunk 3]
```diff
@@ -374,39 +374,53 @@
...
+                                        <b-row>
+                                            <b-col cols="12">
+                                                <editor v-if="rawJson" v-model="userEditJSON" @init="editorInit" lang="json" height="1000"/>
+                                                <b-row v-else-if="userEdit">
                                                     <b-col cols="6">
-                                                        <span class="form-header">Profile</span>
-                                                        <editor v-model="profile" @init="editorInit" lang="json" height="300"/>
-                                                        <!-- <b-form-textarea v-if="userEdit.profile" v-model="userEdit.profile" rows="8"/> -->
+                                                        <h4>Public Profile</h4>
+                                                        <span class="form-header">Position</span>
```
Comment (soichih) [inline ui/src/settings.vue:393]: `position` should only be a private profile (remove it)

File: ui/src/settings.vue [hunk 4]
```diff
@@ -684,11 +793,17 @@ export default {
...
+            Object.keys(this.scopeModel).forEach(entry => {
+                this.userEdit.scopes[entry] = Object.keys(this.scopeModel[entry]).filter(val=>this.scopeModel[entry][val] == true);
+            });
             this.userEdit.ext.openids[0] = this.openids;
             this.$http.put(Vue.config.auth_api+"/user/"+this.userEdit.sub,this.userEdit).then(res=>{
                 this.$notify({type: "success", text: res.data.message});
+                this.showModal = false;
                 this.userEdit = null;
                 this.users = [];
                 this.scopes = null;
```
Comment (soichih) [inline ui/src/settings.vue:809]: Does `this.scopes` need to exist?

File: ui/src/settings.vue [hunk 5]
```diff
@@ -815,14 +930,24 @@ export default {
         },
         tabID() {
             return this.tabs[this.tab].id; 
+        },
+        userScopeEdit() {
```
Comment (soichih) [inline ui/src/settings.vue:934]: What is this function for?

File: ui/src/settings.vue [hunk 6]
```diff
@@ -815,14 +930,24 @@ export default {
...
+            return data;
         }
     },
 
     watch: {
         '$route': function() {
             this.handleRouteParams();
         },
-        tab : function() {
+        tab: function() {
```
Comment (soichih) [inline ui/src/settings.vue:950]: Thanks! :)

File: ui/src/settings.vue [hunk 7]
```diff
@@ -523,6 +545,7 @@ export default {
                 },
                 times: {}
             },
+            selected : null,
```
Comment (soichih) [inline ui/src/settings.vue:548]: What is this used for?

File: ui/src/settings.vue [hunk 8]
```diff
@@ -684,11 +793,17 @@ export default {
...
         submitUser(e) {
             e.preventDefault();
-            this.userEdit.profile = JSON.parse(this.profile||"{}");
-            this.userEdit.scopes = JSON.parse(this.scopes||"{}");
+            if(!this.checkValidJson(this.userEditJSON)) {
+                this.$notify({type: "error", text: "Json formatting Error"});
+                return;
+            }
+            Object.keys(this.scopeModel).forEach(entry => {
+                this.userEdit.scopes[entry] = Object.keys(this.scopeModel[entry]).filter(val=>this.scopeModel[entry][val] == true);
```
Comment (soichih) [inline ui/src/settings.vue:801]: Can scopeModel be merged to userEdit so you don't have to do this?

File: ui/src/settings.vue [hunk 9]
```diff
@@ -835,7 +960,13 @@ export default {
                 });
             }
         },
-        queryUser : function() {
+        userEditJSON: function() {
```
Comment (soichih) [inline ui/src/settings.vue:963]: Please change the function names to reflect what these functions actually do. Maybe something like "convertRawJSONtoUserEdit"?

File: ui/src/settings.vue [hunk 10]
```diff
@@ -684,11 +793,17 @@ export default {
...
+            }
+            Object.keys(this.scopeModel).forEach(entry => {
+                this.userEdit.scopes[entry] = Object.keys(this.scopeModel[entry]).filter(val=>this.scopeModel[entry][val] == true);
+            });
             this.userEdit.ext.openids[0] = this.openids;
             this.$http.put(Vue.config.auth_api+"/user/"+this.userEdit.sub,this.userEdit).then(res=>{
                 this.$notify({type: "success", text: res.data.message});
+                this.showModal = false;
                 this.userEdit = null;
                 this.users = [];
```
Comment (soichih) [inline ui/src/settings.vue:808]: Can't we just update the specific `this.users` so you don't have to invalidate the whole array?

File: ui/src/settings.vue [hunk 11]
```diff
@@ -537,13 +560,55 @@ export default {
...
+                {id: 'profile', label: "Profile"},
+                {id: 'account', label: "Account"},
+                {id: 'notification', label: "Notification"},
+                {id: "users", label: "Users"},
+                {id: "groups", label: "Groups"},
+            ],
+
+            permissionScope: {
+                "brainlife": {
+                    "user" : "description",
```
Comment (soichih) [inline ui/src/settings.vue:591]: for description "Basic user privileges given to all new users by default"

File: ui/src/settings.vue [hunk 12]
```diff
@@ -537,13 +560,55 @@ export default {
...
+                    "admin" : "description of admin"
```
Comment (soichih) [inline ui/src/settings.vue:592]: for description "Allows user to perform administrative tasks for brainlife in general"

File: ui/src/settings.vue [hunk 13]
```diff
@@ -537,13 +560,55 @@ export default {
...
+                },
+                "warehouse": {
+                    "admin" : "description of admin of warehouse",
```
Comment (soichih) [inline ui/src/settings.vue:595]: for description "Allows user to perform administrative tasks specific to warehouse service".

File: ui/src/settings.vue [hunk 14]
```diff
@@ -537,13 +560,55 @@ export default {
...
+                    "datatype.create" : "description of datatype.create in warehouse"
```
Comment (soichih) [inline ui/src/settings.vue:596]: for description "Allows user to register new datatype"

File: ui/src/settings.vue [hunk 15]
```diff
@@ -537,13 +560,55 @@ export default {
...
+                },
+                "amaretti": {
+                    "admin" : "description of admin of amaretti",
```
Comment (soichih) [inline ui/src/settings.vue:599]: for description "Allows user to perform administrative tasks on amaretti service"

File: ui/src/settings.vue [hunk 16]
```diff
@@ -537,13 +560,55 @@ export default {
...
+                    "resource.create" : "description of resource.create in amaretti"
```
Comment (soichih) [inline ui/src/settings.vue:600]: for description "Allows user to register new resource"

File: ui/src/settings.vue [hunk 17]
```diff
@@ -537,13 +560,55 @@ export default {
...
+                },
+                "auth": {
+                    "admin" : "description of admin of auth"
```
Comment (soichih) [inline ui/src/settings.vue:603]: for description "Allows user to perform administrative tasks on authorization service"

**Review by soichih — CHANGES_REQUESTED** [review]

File: ui/src/settings.vue (hunk 9 above)
Comment (bhatiadheeraj) [reply ui/src/settings.vue:963]: It is the model in watcher which updates the UI (userEdit) from the JSON editor (userEditJSON). 
How about I change the variable to userEditJSONtoUI?

**Pull request merged**